### ISSUE TITLE
tsql: DECLARE ... CURSOR FOR with UNION (or any set operation) in the SELECT body fails to parse

### ISSUE THREAD

**Issue by msergein** [sqlfluff/sqlfluff#7809]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

In the `tsql` dialect, a `DECLARE … CURSOR FOR …` whose body is a set-expression
(`SELECT … UNION ALL SELECT …`, or `UNION` / `INTERSECT` / `EXCEPT`) is reported
as an unparsable section. A simple non-set SELECT body parses cleanly; adding a
`UNION ALL` immediately breaks the entire enclosing batch.

T-SQL allows any `<select_statement>` after `FOR` in `DECLARE CURSOR`, including
set operations:

> DECLARE cursor_name CURSOR ... FOR select_statement

where `select_statement` is the full SELECT grammar — set expressions included.

Reference: https://learn.microsoft.com/en-us/sql/t-sql/language-elements/declare-cursor-transact-sql

### Root cause (from reading the source) by claude code

In `src/sqlfluff/dialects/dialect_tsql.py` around L3659:

```python
class DeclareCursorStatementSegment(BaseSegment):
    type = "declare_segment"
    match_grammar = Sequence(
        "DECLARE",
        Ref("CursorNameGrammar"),
        OneOf(
            Ref("CursorDefinitionSegment"),
            Sequence(
                AnySetOf(
                    Ref.keyword("INSENSITIVE", optional=True),
                    Ref.keyword("SCROLL", optional=True),
                ),
                "CURSOR",
                "FOR",
                Ref("SelectStatementSegment"),   # <-- single SELECT only
                ...
            ),
        ),
    )

### Expected Behaviour

The following should parse cleanly under `--dialect tsql`:

```sql
CREATE PROCEDURE [dbo].[p]
AS
BEGIN
    DECLARE cur CURSOR LOCAL FAST_FORWARD FOR
        SELECT 1 AS Id
        UNION ALL
        SELECT 2 AS Id
    OPEN cur
END
```

### Observed Behaviour

```
$ sqlfluff parse --dialect tsql repro.sql
==== parsing violations ====
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'CREATE PROCEDURE
                       | [dbo].[p]\nAS\nBEGIN\n...'
WARNING: Parsing errors found and dialect is set to 'tsql'. Have you configured your dialect correctly?
```

Removing `UNION ALL` and the second `SELECT` makes the file parse cleanly.

### How to reproduce

1. Save the snippet above as `repro.sql`.
2. Run `sqlfluff parse --dialect tsql repro.sql`.
3. Observe the unparsable section starting at line 1.
4. Replace the body with a single `SELECT 1 AS Id` — file parses, no PRS.

### Dialect

tsql

### Version

sqlfluff, version 4.1.0
Python 3.14.4 (Windows)

### Configuration

[sqlfluff]
dialect = tsql

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
